contentcuration/contentcuration/utils/publish.py

```diff
@@ -819,62 +817,74 @@ def fill_published_fields(channel, version_notes):
     channel.save()
 
 
-def create_or_update_tsvectors(channel_id):
+def sync_contentnode_and_channel_tsvectors(channel_id):
     """
-    Create or update tsvectors for the channel and all its content nodes.
+    Creates, deletes and updates tsvectors of the channel and all its content nodes
+    to reflect the current state of channel's main tree.
     """
     # Update or create channel tsvector entry.
     logging.info("Starting to set tsvectors for channel with id {}.".format(channel_id))
 
     from contentcuration.viewsets.channel import primary_token_subquery
 
     channel = (ccmodels.Channel.objects
+               .filter(pk=channel_id)
                .annotate(primary_channel_token=primary_token_subquery,
                          keywords_tsvector=CHANNEL_KEYWORDS_TSVECTOR)
-               .get(pk=channel_id))
+               .values("keywords_tsvector", "main_tree__tree_id")
+               .get())
 
     if ChannelFullTextSearch.objects.filter(channel_id=channel_id).exists():
```

Comment:
Can probably just use `update_or_create` for this instead of the if/else.

Reply:
I discussed this with @bjester and we concluded to use the if else. 

We cannot use `update_or_create` because for `update_or_create` to work, the column needs to have a unique constraint. ForeignKey doesn't have that. 

Changing to OneToOneField will allow us to use `update_or_create` but I think having more than one search entry for a contentnode or channel can be useful is some scenarios...? If you think having more than one search entry for a node or channel won't be helpful then we can go ahead with OneToOneField and use `update_or_create`.

Reply:
To be clear, my comment on this was that `update_or_create` would eliminate this logic. You had mentioned multiple entries for future flexibility, but until that becomes a reality I think it's worth considering just using `update_or_create`. I'm not exactly sure what enhancements that flexibility will allow?

Also, I don't see anything in the [django docs](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#update-or-create) that a unique constraint is required. In addition, you can specify more than one field to compare against so if two or more fields represent a unique record, it seems we could still use `update_or_create`

Reply:
Yeah, I feel it would be fair to enforce uniqueness to the contentnode column in `ContentNodeFullTextSearch` and channel column in `ChannelFullTextSearch` since I too feel there's not much enhancements we can do with multiple search entries. We can always append more data to tsvectors. So let us move forward with OneToOneField.  

From django docs:

> As described above in [get_or_create()](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#django.db.models.query.QuerySet.get_or_create), this method [`update_or_create`] is prone to a race-condition which can result in multiple rows being inserted simultaneously if uniqueness is not enforced at the database level.

Reply:
Resolved on Slack